lsbfdump-jni-rs/native/src/stdin_seek.rs

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// View https://rust-lang.github.io/api-guidelines/naming.html for naming conventions for traits and structs
+use std::io::{self, Cursor, Read, Seek, SeekFrom};
+
+pub struct StdinWithSeek;
+
+impl Read for StdinWithSeek {
+  fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    io::stdin().read(buf)
+  }
+}
+
+impl Seek for StdinWithSeek {
+  fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+    let mut cursor = Cursor::new(Vec::new());
+    cursor.seek(pos)
```

Comment:
I don't understand this Cursor usage. 

Or rather, if I understand this right, this will create a buffer from the input, and then seek to a position in it, filling it with data from the stdin. 

That's undesirable because if the data being streamed in is huge, and the seek position is far in, you'll allocate all that data. 

Just calling read on the stdin in a loop will provide better streaming behavior. You can call that loop in the open-and-seek-input-stream method. Then this whole file is not needed. 

Do I have that right? 

re: rust - I can't figure out from the online doc for Cursor how the Cursor is being associated to the input stream here. Is Cursor a mixin of some sort? I.e., somehow self is bound to the input stream?

Reply:
So according to https://doc.rust-lang.org/std/io/struct.Cursor.html,

> A Cursor wraps an in-memory buffer and provides it with a [Seek](https://doc.rust-lang.org/std/io/trait.Seek.html) implementation.
> Cursors are used with in-memory buffers, anything implementing [AsRef](https://doc.rust-lang.org/std/convert/trait.AsRef.html)<[u8]>, to allow them to implement [Read](https://doc.rust-lang.org/std/io/trait.Read.html) and/or [Write](https://doc.rust-lang.org/std/io/trait.Write.html), allowing these buffers to be used anywhere you might use a reader or writer that does actual I/O."

But I did implement what you mentioned about looping over read on the stdin.

Reply:
Ok. I am learning a lot here. 

So Read and Write are operations on abstract buffers. Cursor makes a buffer have seek. 

And, oh, btw: input streams and output streams are buffers because they support Read and Write. 

Not the same pattern of OO factoring as other languages, but works equally well. Maybe better. In Rust there is no need to construct ByteBufferInputStream that simulates an InputStream using a memory buffer. They've flipped it. Buffer is the primitive here, and I/O Streams are Buffers. 

Works for me.